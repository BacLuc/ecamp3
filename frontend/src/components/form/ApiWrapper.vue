<!--
Displays a field as text or as an input field, depending on the editing prop.
You can two-way bind to the value using v-model.
-->

<template>
  <span>
    <span v-if="!editing">{{ value }}</span>
    <span v-if="editing">
      <v-form
        inline
        class="mb-2">
        <slot :editing="editing"
              :localValue="localValue"
              :errorMessage="errorMessage"
              :isSaving="isSaving"
              :status="status"
              :on="eventHandlers" />

        <v-btn
          v-if="!autoSave"
          small
          color="warning"
          class="mb-1"
          @click="reset">
          Reset
        </v-btn>

        <v-btn
          v-if="!autoSave"
          color="primary"
          small
          :disabled="isSaving || (required && $v.localValue.$invalid)"
          class="mr-2 ml-2 mb-1"
          @click="save">
          <v-progress-circular
            v-if="isSaving"
            indeterminate
            color="primary"
            size="20"
            class="mr-2" />

          <v-icon
            v-if="showSuccessIcon"
            left>
            mdi-check
          </v-icon>

          Save
        </v-btn>
      </v-form>
    </span>
  </span>
</template>

<script>

import { debounce } from 'lodash'
import { validationMixin } from 'vuelidate'
import { required } from 'vuelidate/lib/validators'
import { apiPropsMixin } from '@/mixins/apiPropsMixin'

export default {
  name: 'ApiWrapper',
  mixins: [apiPropsMixin, validationMixin],
  data () {
    return {
      localValue: this.value,
      isSaving: false,
      showSuccessIcon: false,
      dirty: false,
      eventHandlers: {
        save: this.save,
        reset: this.reset,
        input: this.onInput,
        touch: this.touch
      }
    }
  },
  validations: {
    localValue: {
      required
    }
  },
  computed: {
    isDirty: function () {
      return this.value !== this.localValue
    },
    errorMessage () {
      const errors = []
      if (!this.$v.localValue.$dirty) return errors
      !this.$v.localValue.required && errors.push('Feld darf nicht leer sein.')
      return errors
    },
    status: function () {
      if (this.isSaving) {
        return 'saving'
      } else if (this.showSuccessIcon) {
        return 'success'
      } else {
        return 'init'
      }
    }
  },
  watch: {
    value: function (newValue, oldValue) {
      // override local value if it wasn't dirty
      if (!this.dirty || this.overrideDirty) {
        this.localValue = newValue
      }

      // clear dirty if outside value changes to same as local value (e.g. after save operation)
      if (this.localValue === newValue) {
        this.dirty = false
      }
    }
  },
  created: function () {
    // Create debounced save method (lodash debounce function)
    this.debouncedSave = debounce(this.save, this.autoSaveDelay)
  },
  methods: {

    log: function (title, event) {
      console.log(title)
      console.log(event)
    },
    touch: function () {
      this.$v.localValue.$touch()
    },
    onInput: function (newValue) {
      this.localValue = newValue
      this.dirty = true
      this.$v.localValue.$touch()

      if (this.autoSave) {
        this.debouncedSave()
      }
    },
    reset: function (event) {
      this.localValue = this.value
      this.$v.localValue.$reset()
    },
    save: function (event) {
      this.$v.localValue.$touch()
      if (this.required && this.$v.localValue.$anyError) {
        return
      }

      this.isSaving = true
      this.api.patch(this.uri, { [this.fieldname]: this.localValue }).then(() => {
        this.isSaving = false
        this.$v.localValue.$reset()

        this.showSuccessIcon = true
        setTimeout(() => { this.showSuccessIcon = false }, 2000)
      }, (e, a) => {
        console.log('error')
      })
    }
  }

}
</script>

<style scoped>
</style>
