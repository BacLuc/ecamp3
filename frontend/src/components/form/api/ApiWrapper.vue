<!--
Wrapper component for form components to save data back to API
-->

<template>
  <v-form
    :class="['d-flex','flex-wrap',{'api-wrapper--inline':!autoSave && !readonly && !separateButtons, 'my-4':!noMargin}]"
    @submit.prevent="onEnter">
    <slot
      :localValue="localValue"
      :errorMessages="errorMessages"
      :isSaving="isSaving"
      :status="status"
      :on="eventHandlers" />

    <div
      v-if="!autoSave && !readonly"
      :class="['d-flex', {'my-1 ml-auto': separateButtons}]">
      <v-btn
        :disabled="disabled"
        small
        elevation="0"
        :class="{'mr-1': separateButtons}"
        :tile="!separateButtons"
        :height="separateButtons ? '' : 'auto'"
        @click="reset">
        Reset
      </v-btn>

      <v-btn
        color="primary"
        small
        elevation="0"
        :disabled="disabled || (required && $v.localValue.$invalid)"
        class="v-btn--last-instance"
        :height="separateButtons ? '' : 'auto'"
        :loading="isSaving"
        @click="save">
        Save
      </v-btn>
    </div>
  </v-form>
</template>

<script>

import { debounce } from 'lodash'
import { validationMixin } from 'vuelidate'
import { required } from 'vuelidate/lib/validators'
import { apiPropsMixin } from '@/mixins/apiPropsMixin'

export default {
  name: 'ApiWrapper',
  mixins: [apiPropsMixin, validationMixin],
  props: {
    separateButtons: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      localValue: null,
      isSaving: false,
      showIconSuccess: false,
      dirty: false,
      eventHandlers: {
        save: this.save,
        reset: this.reset,
        input: this.onInput,
        touch: this.touch
      }
    }
  },
  validations: {
    localValue: {
      required
    }
  },
  computed: {
    isDirty: function () {
      return this.apiValue !== this.localValue
    },
    errorMessages () {
      const errors = []
      if (!this.$v.localValue.$dirty) return errors
      !this.$v.localValue.required && errors.push('Feld darf nicht leer sein.')
      return errors
    },
    status: function () {
      if (this.isSaving) {
        return 'saving'
      } else if (this.showIconSuccess) {
        return 'success'
      } else {
        return 'init'
      }
    },
    debouncedSave () {
      return debounce(this.save, this.autoSaveDelay)
    },
    apiValue () {
      // return value from API unless `value` is set explicitly
      return this.value || this.api.get(this.uri)[this.fieldname]
    }
  },
  watch: {
    apiValue: function (newValue, oldValue) {
      // override local value if it wasn't dirty
      if (!this.dirty || this.overrideDirty) {
        this.localValue = newValue
      }

      // clear dirty if outside value changes to same as local value (e.g. after save operation)
      if (this.localValue === newValue) {
        this.dirty = false
      }
    }
  },
  created () {
    // initial data load from API
    this.localValue = this.apiValue
  },
  methods: {
    touch () {
      this.$v.localValue.$touch()
    },
    onInput: function (newValue) {
      this.localValue = newValue
      this.dirty = true
      this.touch()

      if (this.autoSave) {
        this.debouncedSave()
      }
    },
    reset () {
      this.localValue = this.apiValue
      this.$v.localValue.$reset()
    },
    onEnter () {
      if (!this.autoSave) {
        this.save()
      }
    },
    save () {
      // abort saving if component is in readonly or disabled state
      // this is here for safety reasons, should not be triggered if the wrapped component behaves normally
      if (this.readonly || this.disabled) {
        return
      }

      // abort saving in case of validation errors
      this.touch()
      if (this.$v.localValue.$anyError) {
        return
      }

      // reset all dirty flags and start saving
      this.dirty = false
      this.$v.localValue.$reset()
      this.isSaving = true

      this.api.patch(this.uri, { [this.fieldname]: this.localValue }).then(() => {
        this.isSaving = false
        this.showIconSuccess = true
        setTimeout(() => {
          this.showIconSuccess = false
        }, 2000)
      }, (e, a) => {
        console.log('error')
      })
    }
  }

}
</script>

<style lang="scss" scoped>
  .api-wrapper--inline .v-btn--last-instance {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
  .api-wrapper--inline .v-btn {
    border-top: 1px solid rgba(0, 0, 0, 0.38);
    border-bottom: 1px solid rgba(0, 0, 0, 0.38);
  }
</style>

<style lang="scss" >
  .api-wrapper--inline .v-text-field {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
</style>
