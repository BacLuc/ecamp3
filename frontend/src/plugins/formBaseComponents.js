// Global registration of all Vue components in folder components/form/base

// import upperFirst from 'lodash/upperFirst'
// import camelCase from 'lodash/camelCase'

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

/*
const requireComponent = require.context(
  // The relative path of the components folder
  '../components/form/base',
  // Whether or not to look in subfolders
  false,
  // The regular expression used to match base component filenames
  /[A-Z]\w+\.(vue|js)$/
) */

class FormBaseComponentsPlugin {
  install (Vue, options) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)

    /*
    requireComponent.keys().forEach(fileName => {
      // Get component config
      const componentConfig = requireComponent(fileName)

      // Get PascalCase name of component
      const componentName = upperFirst(
        camelCase(
          // Gets the file name regardless of folder depth
          fileName
            .split('/')
            .pop()
            .replace(/\.\w+$/, '')
        )
      )

      // Register component globally
      Vue.component(
        componentName,
        // Look for the component options on `.default`, which will
        // exist if the component was exported with `export default`,
        // otherwise fall back to module's root.
        componentConfig.default || componentConfig
      )
    }) */
  }
}

export default new FormBaseComponentsPlugin()
