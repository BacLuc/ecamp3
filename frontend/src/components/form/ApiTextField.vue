<!--
Displays a field as text or as an input field, depending on the editing prop.
You can two-way bind to the value using v-model.
-->

<template>
  <api-wrapper
    v-bind="$props">
    <template slot-scope="wrapper">
      <v-text-field
        :value="wrapper.localValue"
        v-bind="$attrs"
        :label="label"
        :error-messages="wrapper.errorMessage"
        @input="wrapper.on.input"
        @blur="wrapper.on.touch">
        <status-icon slot="append" :status="wrapper.status" />
      </v-text-field>
    </template>
  </api-wrapper>
</template>

<script>
import { apiPropsMixin } from '@/mixins/apiPropsMixin'
import ApiWrapper from './ApiWrapper'
import StatusIcon from './StatusIcon'

export default {
  name: 'ApiTextField',
  components: { ApiWrapper, StatusIcon },
  mixins: [apiPropsMixin],

  data () {
    return {}
  }
}
</script>

<style scoped>
</style>
