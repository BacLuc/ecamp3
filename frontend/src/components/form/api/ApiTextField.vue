<!--
Displays a field as a e-text-field + write access via API wrapper
-->

<template>
  <api-wrapper
    v-slot="wrapper"
    v-bind="$props">
    <e-text-field
      :value="wrapper.localValue"
      v-bind="$attrs"
      :readonly="readonly"
      :disabled="disabled"
      :error-messages="wrapper.errorMessages"
      :loading="wrapper.isSaving ? 'secondary' : false"
      no-margin
      outlined
      :filled="false"
      @input="wrapper.on.input"
      @blur="wrapper.on.touch">
      <template #append>
        <icon-success :visible="wrapper.status === 'success'" />
      </template>
    </e-text-field>
  </api-wrapper>
</template>

<script>
import { apiPropsMixin } from '@/mixins/apiPropsMixin'
import ApiWrapper from './ApiWrapper'
import IconSuccess from './IconSuccess'

export default {
  name: 'ApiTextField',
  components: { ApiWrapper, IconSuccess },
  mixins: [apiPropsMixin],

  data () {
    return {}
  }
}
</script>

<style scoped>
</style>
