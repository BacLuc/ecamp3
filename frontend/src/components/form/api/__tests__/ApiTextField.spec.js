// Libraries
import Vue from 'vue'
import Vuetify from 'vuetify'
import flushPromises from 'flush-promises'

import formBaseComponents from '@/plugins/formBaseComponents'

import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import ApiTextField from '../ApiTextField.vue'
import ApiWrapper from '../ApiWrapper'
import { ValidationObserver } from 'vee-validate'

jest.mock('lodash')
const { cloneDeep } = jest.requireActual('lodash')

jest.useFakeTimers()
Vue.use(Vuetify)
Vue.use(formBaseComponents)

let vuetify

// config factory
function createConfig (overrides) {
  const mocks = {
    api: {
      patch: () => Promise.resolve()
    }
  }
  const propsData = {
    value: 'Test Value',
    fieldname: 'test-field',
    uri: '/test-field/123',
    label: 'Test Field'
  }
  const stubs = {
    ApiWrapper,
    ValidationObserver
  }
  const localVue = createLocalVue()

  return cloneDeep(Object.assign({ mocks, propsData, stubs, vuetify, localVue }, overrides))
}

describe('ApiTextField.vue', () => {
  beforeEach(() => {
    vuetify = new Vuetify()
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  // keep this the first test --> otherwise element IDs change constantly
  test('renders correctly', () => {
    const config = createConfig()
    const wrapper = shallowMount(ApiTextField, config)

    expect(wrapper.element).toMatchSnapshot()
  })

  test('input change triggers api.patch call and status update', async () => {
    const config = createConfig()
    const patchSpy = jest.spyOn(config.mocks.api, 'patch')
    const wrapper = mount(ApiTextField, config)

    const newValue = 'new value'

    // contains 1 e-text-field
    expect(wrapper.findComponent({ name: 'ETextField' }).exists()).toBe(true)

    wrapper.find('input').setValue(newValue)

    // resolve lodash debounced
    jest.runAllTimers()

    // await validation Promise & api.patch Promise
    await flushPromises()

    expect(patchSpy).toBeCalledTimes(1)
    expect(patchSpy).toBeCalledWith(config.propsData.uri, { [config.propsData.fieldname]: newValue })
  })
})
