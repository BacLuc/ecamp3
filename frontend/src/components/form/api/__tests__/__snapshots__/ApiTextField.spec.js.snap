// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiTextField.vue renders correctly 1`] = `
<v-form-stub
  class="my-4"
>
  <e-text-field-stub
    backgroundcolor=""
    clearicon="$clear"
    errorcount="1"
    errormessages=""
    hidedetails="auto"
    label="Test Field"
    loaderheight="2"
    messages=""
    nomargin="true"
    outlined="true"
    rules=""
    successmessages=""
    type="text"
    value="Test Value"
  />
   
  <!---->
</v-form-stub>
`;
